cmd/incusd/main_forkbpf.go

```diff
@@ -0,0 +1,284 @@
+package main
+
+/*
+#include "config.h"
+
+#include <errno.h>
+#include <fcntl.h>
+#include <sched.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+#include <unistd.h>
+
+#include "incus.h"
+#include "memory_utils.h"
+
+pid_t incus_forkbpf_pid;
+int incus_forkbpf_socket_parent;
+int incus_forkbpf_socket_child;
+
+static int dosetns_file(char *file)
+{
+	__do_close int ns_fd = -EBADF;
+
+	ns_fd = open(file, O_RDONLY);
+	if (ns_fd < 0) {
+		fprintf(stderr, "%m - Failed to open \"%s\": %s", file, strerror(errno));
+		return -1;
+	}
+
+	if (setns(ns_fd, 0) < 0) {
+		fprintf(stderr, "%m - Failed to attach to namespace \"%s\": %s", file, strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+void forkbpf(void)
+{
+	char *pidstr;
+	char path[PATH_MAX];
+	int fds[2];
+
+	pidstr = getenv("LXC_PID");
+	if (!pidstr) {
+		fprintf(stderr, "No LXC_PID in environment\n");
+		_exit(EXIT_FAILURE);
+	}
+
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, fds) < 0) {
+		fprintf(stderr, "Failed to create socket pair: %s", strerror(errno));
+		_exit(EXIT_FAILURE);
+	}
+
+	incus_forkbpf_socket_parent = fds[0];
+	incus_forkbpf_socket_child = fds[1];
+
+	incus_forkbpf_pid = fork();
+	if (incus_forkbpf_pid < 0) {
+		fprintf(stderr, "%s - Failed to create new process\n",
+			strerror(errno));
+		_exit(EXIT_FAILURE);
+	}
+
+	if (incus_forkbpf_pid == 0) {
+		// Attach to the user namespace.
+		snprintf(path, sizeof(path), "/proc/%s/ns/user", pidstr);
+		if (dosetns_file(path) < 0) {
+			_exit(EXIT_FAILURE);
+		}
+
+		// Attach to the user namespace.
+		snprintf(path, sizeof(path), "/proc/%s/ns/mnt", pidstr);
+		if (dosetns_file(path) < 0) {
+			_exit(EXIT_FAILURE);
+		}
```

Comment:
I guess you probably want to use `pidfd_nsfd`+`change_namespaces`, as well as pidfds if possible (just like `forkproxy` does)?